src/OrbitGl/ThreadTrack.cpp

```diff
@@ -436,8 +429,11 @@ void ThreadTrack::DoUpdatePrimitives(PrimitiveAssembler& primitive_assembler,
               CreatePickingUserData(primitive_assembler, *timer_info));
         }
       } else {
-        primitive_assembler.AddVerticalLine(pos, box_height, draw_data.z, color,
-                                            std::move(user_data));
+        // In this case size should be equal to 1. We will use the middle point of the start and
+        // end, so there are no empty spaces when we combine boxes with lines.
+        const float kHalfSize = 0.5f;
```

Comment:
I'd reuse the `kHalfPixel` constant above. I'm not sure about this though - you're explicitly moving the line in between two pixels, this could result in the line being drawn as 2 pixels, or randomly in of of both, or maybe even disappear (like we've seen with the horizontal lines)? I don't know if the gl specification talks about this case and it is well-defined, otherwise it might be up to the driver to decide. This sounds like a hack that hides an issue.

Are you sure this is needed?

Reply:
I completely agree that it's a hack hiding an issue because it doesn't make too much sense but without this, the issue was still there. I can add a TODO here. But I tried for several hours and this was the only combination that worked for me. As I said in the body, the other approach is just drawing rectangles and we don't need any half pixel.

Reply:
It works for *you*, that's what I'm concerned about. OpenGl doesn't guarantee this.

Some interesting discussion in this thread: https://stackoverflow.com/questions/45785957/opengl-line-rasterization-with-integer-endpoints - if you want to dive into the OpenGl specification: It doesn't exactly prescribe how lines need to be rendered, and they explicitly do not guarantee pixel-perfect rendering. The corner case of a line exactly in the middle of two pixels is treated by slightly overlapping the endpoints of the diamonds that the algorithm intersects with per pixel, but the way I read it this not a "must" and depends on GPU and driver implementation. They allow an error of one pixel from the reference implementation, and by specifying the line in the middle of pixels you're basically putting yourself at the mercy of the driver.

Maybe we actually want to take the 10% performance hit instead and not use lines here at all?

Reply:
Reverted. Now we are only drawing boxes in ThreadTrack. I'm open to opinions.